keps/sig-node/20200619-provisioning-resources-with-numa-topology.md

```diff
@@ -0,0 +1,240 @@
+substracting resources of kube cgroup and system cgroup from
+system allocatable resources.
+
+### Design based on CRI
+
+The containerStatusResponse returned as a response to the ContainerStatus rpc contains `Info` field which is used by the container runtime for capturing ContainerInfo.
+```go
+message ContainerStatusResponse {
+      ContainerStatus status = 1;
+      map<string, string> info = 2;
+}
+```
+
+Containerd has been used as the container runtime in the initial investigation. The internal container object info
+[here](https://github.com/containerd/cri/blob/master/pkg/server/container_status.go#L130)
+
+The Daemon set is responsible for the following:
+
+- Parsing the info field to obtain container resource information
+- Identifying NUMA nodes of the allocated resources
+- Identifying total number of resources allocated on a NUMA node basis
+- Detecting Node resource capacity on a NUMA node basis
+- Updating the CRD instance per node indicating available resources on NUMA nodes, which is referred to the scheduler
+
+
+#### Drawbacks
+
+The content of the `info` field is free form, unregulated by the API contract. So, CRI-compliant container runtime engines are not required to add any configuration-specific information, like for example cpu allocation, here. In case of containerd container runtime, the Linux Container Configuration is added in the `info` map depending on the verbosity setting of the container runtime engine.
+
+There is currently work going on in the community as part of the the Vertical Pod Autoscaling feature to update the ContainerStatus field to report back containerResources
+[KEP](https://github.com/kubernetes/enhancements/blob/master/keps/sig-node/20191025-kubelet-container-resources-cri-api-changes.md).
+
+
+### Design based on podresources interface of the kubelet
+
+Podresources interface of kubelet described in
+
+pkg/kubelet/apis/podresources/v1alpha1/api.proto
+
+it available for every process on the worker node by
+unix domain socket situated by the following path:
+
+filepath.Join(kl.getRootDir(), config.DefaultKubeletPodResourcesDirName)
+
+it could be used to collect used resources on the worker node and to evaluate
+its NUMA assignment (by device id)
+
+For example
+TODO add path on the real system
+
+Currently it doesn't provide information from CPUManager only from
+DeviceManager as well as information about memory which was used.
+But information provided by podresource could be easily extended
+without modification of API.
+
+### API
+
+Available resources with topology of the node should be stored in CRD. Format of the topology described
+[in this document](https://docs.google.com/document/d/12kj3fK8boNuPNqob6F_pPU9ZTaNEnPGaXEooW1Cilwg/edit).
+
+
+```go
+// NodeResourceTopology is a specification for a Foo resource
+type NodeResourceTopology struct {
+       metav1.TypeMeta   `json:",inline"`
+       metav1.ObjectMeta `json:"metadata,omitempty"`
+
+       TopologyPolicy string
```

Comment:
How is this daemon going to obtain the node's topology policy?

Reply:
This daemon should have access to KubletConfiguration. If it in the file system it should be able to read it. Dynamic Kubelet Configuration I think also will be supported.

Reply:
What about flags?

Reply:
Handling of flags could be useful especially for --reserved-cpus, which CPUs are reserved for the kubelet and which can be used for the workload. Workloads that need pinning could ask the scheduler for lets say 3 isolated/pinnable/dedicates CPUs, which are needed e.g. for things related to low latency or high throughput.

Reply:
In case of command line option nfd-worker should have access to procfs. It looks not so graceful (requires more permissions). Maybe in this case extend planned GetAvailableDevices to GetAvailableResources and provide list of cpu id, it can help to avoid dealing with --reserved-cpus, system & kube cgroups.